File: nodejs/patterns/src/creational/object_pool.js
class Worker {
  constructor(workerId) {
    this.workerId = workerId;
    this.busy = false;
  }

  reset() {
    this.busy = false;
  }
}

class WorkerPool {
  constructor(size) {
    this.available = Array.from({ length: size }, (_, index) => new Worker(index));
    this.inUse = new Set();
  }

  acquire() {
    if (this.available.length === 0) {
      throw new Error("pool exhausted");
    }
    const worker = this.available.pop();
    worker.busy = true;
    this.inUse.add(worker.workerId);
    return worker;
  }

  release(worker) {
    if (!this.inUse.has(worker.workerId)) {
      throw new Error("worker was not checked out");
    }
    this.inUse.delete(worker.workerId);
    worker.reset();
    this.available.push(worker);
  }
}

function demo() {
  const pool = new WorkerPool(2);
  const first = pool.acquire();
  pool.acquire();
  pool.release(first);
  const again = pool.acquire();
  return first.workerId === again.workerId;
}

module.exports = { WorkerPool, demo };

File: python/patterns/src/prototype_patterns/creational/object_pool.py
"""Object Pool pattern example."""

from __future__ import annotations


class Worker:
    def __init__(self, worker_id: int) -> None:
        self.worker_id = worker_id
        self.busy = False

    def reset(self) -> None:
        self.busy = False


class WorkerPool:
    def __init__(self, size: int) -> None:
        self._available = [Worker(index) for index in range(size)]
        self._in_use: set[int] = set()

    def acquire(self) -> Worker:
        if not self._available:
            raise RuntimeError("pool exhausted")
        worker = self._available.pop()
        worker.busy = True
        self._in_use.add(worker.worker_id)
        return worker

    def release(self, worker: Worker) -> None:
        if worker.worker_id not in self._in_use:
            raise RuntimeError("worker was not checked out")
        self._in_use.remove(worker.worker_id)
        worker.reset()
        self._available.append(worker)


def demo() -> bool:
    pool = WorkerPool(2)
    first = pool.acquire()
    second = pool.acquire()
    pool.release(first)
    again = pool.acquire()
    return first.worker_id == again.worker_id and second.busy
